### PR TITLE
feat(lsp): use `g:markdown_fenced_languages` in `vim.lsp.util.stylized_markdown`

### LINKED ISSUE #14616: `ts` not recognized for syntax highlights in the fancy_floating_markdown popup
- `nvim --version`: v0.5.0-dev+1345-g6dd04ed5f
- Operating system/version: Centos 8 Linux
- Terminal name/version: Gnome Terminal
- `$TERM`: xterm-256color

When using the LSP hover typescript code blocks are not syntax highlighted when ` ```ts` is used, but only when  ` ```typescript` is. This means that any language server using ` ```ts` fails to render any syntax highlighting. `ts` is the most commonly used.

In my `init.vim` I have:
```
let g:markdown_fenced_languages = [ 'typescript', 'ts=typescript' ]
```
Which works fine for Markdown files using ` ```ts`:

Screenshot from 2021-05-22 14-52-19

But in the LSP hover for Deno that uses ` ```ts` code blocks I get no highlighting:
Screenshot from 2021-05-22 08-44-30

### PR DESCRIPTION
**Pull request by folke**
Markdown text coming from `hover` or `signature_help` or other sources, that contains code blocks are processed in `fancy_floating_markdown` and replaced by syntax regions.

Currently the code block `lang` is used as the `syntax`, but this is not always correct.

Vim already has support for mapping code block langs to syntaxes with `g:markdown_fenced_languages`.

This PR simply uses that variable to provide better mappings from langs to syntax. (fixes #14616)

Without this PR:
image

With this PR:
image

Example config:

```lua
-- Use proper syntax highlighting in code blocks
local fences = {
  "lua",
  "vim",
  "json",
  "typescript",
  "javascript",
  "js=javascript",
  "ts=typescript",
  "shell=sh",
  "python",
  "sh",
  "bash=sh",
  "console=sh",
}
vim.g.markdown_fenced_languages = fences
```

### CODE REVIEW

**Comment by folke** [comment]
I wonder if we maybe should also add some extra defaults for common languages, like:
* tsx=typescriptreact
* jsx=javascriptreact
* js=javascript
* ts=typescript
* shell=sh
* bash=sh
* console=sh

WDYT?

**Comment by David-Else** [comment]
> I wonder if we maybe should also add some extra defaults for common languages, like:
> 
>     * tsx=typescriptreact
> 
>     * jsx=javascriptreact
> 
>     * js=javascript
> 
>     * ts=typescript
> 
>     * shell=sh
> 
>     * bash=sh
> 
> 
> WDYT?

It would make sense I think, and save a lot of bother as many people won't use Markdown anywhere else and won't know or care about `g:markdown_fenced_languages`. It would be best if the user's `g:markdown_fenced_languages` overrides any of the defaults, in case they have any strange special use cases.

**Comment by mjlbach** [comment]
FWIW I'm not sure we should add defaults for markdown fenced languages globally, as this will affect the default markdown highlighting behavior. We *could* add this in lsp_markdown.vim which inherits from markdown.vim (which is what reads markdown_fenced_languages IIRC).

**Comment by folke** [comment]
Yeah, definitely not add that to the global list.

> We could add this in lsp_markdown.vim which inherits from markdown.vim (which is what reads markdown_fenced_languages IIRC).

We mostly need the fences inside the `fancy_floating_markdown` where  we create the code block syntax regions. Exposing the defults further to `lsp_markdown` is a plus, but less important I guess.

I would just have a small table with default mappings in `lsp.util`. For additional mappings, the user can then configure the `markdown_fenced_languages`.

What would you suggest on how to do this?

Another option could be that we have a default list inside util, that users can change if needed. So not doing anything then with `g:markdown_fenced_languages`. It would be an even simpler change and would probably make more sense?

So basically inside `vim.lsp.util`:

```lua
-- table of `lang` to `syntax`
M.markdown_fenced_syntax = {
  ts = "typescript",
  js = "javascript",
...
}
```

**Comment by mjlbach** [comment]
Ah, I was saying we could use lspmarkdown to set buffer local variables which could then be overridden by the user, so setting the markdown_fenced_languages locally for anything that invokes the lsp_markdown filetype etc.

I think I'm cool with adding a default list with the common mappings, mildly concerned this might get large though.

**Comment by folke** [comment]
I think I still don't fully get what you mean 😅

Just to make sure we're talking about the same thing here:

* `fancy_floating_markdown` is the **only** place that uses the `lsp_markdown` (right now at least)
* `lsp_markdown` (and the included `markdown` syntax) **never** renders any code blocks, since those are taken care of inside `fancy_floating_markdown`, using syntax regions. We strip the ` ``` ` parts.
* so `g: markdown_fenced_languages` would never be used in `lsp_markdown`, since there are no code blocks

> I think I'm cool with adding a default list with the common mappings (not enabled by default though)

Right now in  `fancy_floating_markdown` we're trying to render every code block using the correct syntax region. What would be the reason to make the default list not enabled by default?

Basically if we find a code block with lang `ts`, we know that it should be rendered with the `typescript` syntax.

> mildly concerned this might get large though.
that's definitely a risk. We could try to keep it small. for most languages they are probably already the same as the syntax name. Users can always extend it with their own stuff if they need to.

**Comment by mjlbach** [comment]
> lsp_markdown (and the included markdown syntax) never renders any code blocks, since those are taken care of inside fancy_floating_markdown, using syntax regions. We strip the ``` parts.

Right, we're currently using only the syntax part of lsp_markdown. What I'm saying, is we can add buffer local variables in the filetype to automatically set things like markdown_fenced_languages as buffer local variables on all of the lsp related hover. Just a suggestion.

> What would be the reason to make the default list not enabled by default?

If we do not override markdown_fenced_languages (as was originally brought up), I think it can be a default for the hover windows.

**Comment by folke** [comment]
Got it, thanks!

Just made some small final changes. Should be ready now.

**Comment by folke** [comment]
I'll add this to the config PR https://github.com/neovim/neovim/pull/14681. Makes more sense since it's configuration.

**Comment by folke** [comment]
Superseded by #14681

**Comment by folke** [comment]
@mjlbach I reopened this PR with only the minimal needed to be able to have syntax highlighting for code blocks where a lang is not the same as a syntax.

The new change only checks in `markdown_fenced_languages` and uses any mappings there to apply the correct syntax regions.

This is completely inline with how the regular markdown syntax renders code blocks.

**Review by mjlbach — APPROVED** [review]
